crates/ignore/src/types.rs

```diff
@@ -112,35 +114,241 @@ use crate::{Error, Match};
 /// The lifetime `'a` refers to the lifetime of the underlying file type
 /// definition, which corresponds to the lifetime of the file type matcher.
 #[derive(Clone, Debug)]
-pub struct Glob<'a>(GlobInner<'a>);
+pub struct Globs(GlobsInner);
 
 #[derive(Clone, Debug)]
-enum GlobInner<'a> {
+enum GlobsInner {
     /// No glob matched, but the file path should still be ignored.
     UnmatchedIgnore,
     /// A glob matched.
     Matched {
         /// The file type definition which provided the glob.
-        def: &'a FileTypeDef,
+        file_type_indices_token: FileTypeIndicesToken,
     },
 }
 
-impl<'a> Glob<'a> {
-    fn unmatched() -> Glob<'a> {
-        Glob(GlobInner::UnmatchedIgnore)
+impl Globs {
+    pub fn match_metadata_token(&self) -> Option<MatchMetadataToken> {
+        match &self.0 {
+            GlobsInner::Matched { file_type_indices_token } => {
+                Some(MatchMetadataToken::FileTypeIndicesToken(
+                    file_type_indices_token.clone(),
+                ))
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum MatchMetadataToken {
+    FileTypeIndicesToken(FileTypeIndicesToken),
+}
+
+const FILE_TYPE_INDICES_BIT_ARRAY_LEN: usize = 256;
+
+type FileTypeIndicesBitArray = BitArr!(for FILE_TYPE_INDICES_BIT_ARRAY_LEN);
+
+fn bit_array_into_file_type_indices_iterator(
+    bit_array: FileTypeIndicesBitArray,
+) -> FileTypeIndicesBitArrayIterator {
+    FileTypeIndicesBitArrayIterator::new(bit_array)
+}
+
+pub struct FileTypeIndicesBitArrayIterator {
+    bit_array_iterator:
+        Enumerate<<FileTypeIndicesBitArray as IntoIterator>::IntoIter>,
+}
+
+impl FileTypeIndicesBitArrayIterator {
+    pub fn new(bit_array: FileTypeIndicesBitArray) -> Self {
+        Self { bit_array_iterator: bit_array.into_iter().enumerate() }
+    }
+}
+
+impl Iterator for FileTypeIndicesBitArrayIterator {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some((file_type_index, is_selected)) =
+            self.bit_array_iterator.next()
+        {
+            if is_selected {
+                return Some(file_type_index);
+            }
+        }
+        None
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum FileTypeIndicesToken {
+    Empty,
```

Comment:
This is maybe overkill and could just be the `BitArray`/`Vec` variants (ie how much computation or whatever am I saving by special-casing `Empty`/`Single`)?